### ISSUE TITLE
avoid deep-merge of rule config

### ISSUE THREAD

**Issue by xjamundx** [eslint/eslint#3175]
Recently in #2963 we implemented a deep-merge of nested configs. Rule configs probably shouldn't be deep-merged though. Here's what I mean.

.eslintrc
```js
{
    "rules":  {
        "consumerweb/param-names": [2, { "err": ["error", "e"], "callback": ["cb"] }],
    }
}
```

public/js/.eslintrc
```js
"consumerweb/param-names": [2, { "event": ["evt", "e"], "callback": ["cb"] }]
```

#### Result

The config sent to the rule is this incompatible combination:
```js
"consumerweb/param-names": [2, { "error": ["err", "e"], "event": ["evt", "e"], "callback": ["cb"] }]
```
#### Expected

I expected the config in `public/js` to overwrite the earlier config, because merging in this case produces an incompatible configuration.

```js
"consumerweb/param-names": [2, { "event": ["evt", "e"], "callback": ["cb"] }]
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Agreed.

**Comment by xjamundx** [comment]
A good one to test might be `no-restricted-modules` which appears to be getting merged with some odd effects.

**Comment by gyandeeps** [comment]
I already have the solution for this, I will send the PR tomorrow.
